core/lib/zksync_core/src/sync_layer/gossip/mod.rs

```diff
@@ -0,0 +1,99 @@
+//! Consensus adapter for EN synchronization logic.
+
+use anyhow::Context as _;
+use tokio::sync::watch;
+
+use std::sync::Arc;
+
+use zksync_concurrency::{ctx, scope};
+use zksync_consensus_executor::{Executor, ExecutorConfig};
+use zksync_consensus_roles::node;
+use zksync_dal::ConnectionPool;
+
+mod buffered;
+mod conversions;
+mod metrics;
+mod storage;
+#[cfg(test)]
+mod tests;
+mod utils;
+
+use self::{buffered::Buffered, storage::PostgresBlockStorage};
+use super::{fetcher::FetcherCursor, sync_action::ActionQueueSender};
+
+/// Starts fetching L2 blocks using peer-to-peer gossip network.
+pub async fn start_gossip_fetcher(
+    pool: ConnectionPool,
+    actions: ActionQueueSender,
+    executor_config: ExecutorConfig,
+    node_key: node::SecretKey,
+    mut stop_receiver: watch::Receiver<bool>,
+) -> anyhow::Result<()> {
+    let result = scope::run!(&ctx::root(), |ctx, s| {
+        s.spawn_bg(async {
+            if stop_receiver.changed().await.is_err() {
+                tracing::warn!(
+                    "Stop signal sender for gossip fetcher was dropped without sending a signal"
+                );
+            }
+            s.cancel();
+            tracing::info!("Stop signal received, gossip fetcher is shutting down");
+            Ok(())
+        });
+        start_gossip_fetcher_inner(ctx, pool, actions, executor_config, node_key)
+    })
+    .await;
+
+    result.or_else(|err| {
+        if err.root_cause().is::<ctx::Canceled>() {
```

Comment:
can we avoid inspecting the error? I.e. either catch cancellation internally or make the scope::run! return Canceled rather than anyhow (or some other strongly typed error).